### PR TITLE
V0.2 PG, K-step and lambda return utils 

### PR DESCRIPTION
**Pull request by ysqyang**
# Description

## Linked issue(s)/Pull request(s)

- [Issue 129](https://github.com/microsoft/maro/issues/129)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [x] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [x] Add/update the related tests
- [x] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/actor/simple_actor.py [hunk 1]
```diff
@@ -41,7 +42,7 @@ def roll_out(self, model_dict: dict = None, epsilon_dict: dict = None, done: boo
 
         # load models
         if model_dict is not None:
-            self._inference_agents.load_models(model_dict)
+            self._inference_agents.load_trainable_models(model_dict)
```
Comment (ArthurJiang) [inline maro/rl/actor/simple_actor.py:45]: keep the old name

File: maro/rl/agent/abs_agent.py [hunk 2]
```diff
@@ -25,13 +23,10 @@ class AbsAgent(ABC):
...
-        experience_pool (AbsStore): A data store that stores experiences generated by the experience shaper.
+        experience_pool (AbsStore): A data store that stores experiences generated by the experience shaper. This is
+            only necessary for some algorithms. Defaults to None.
     """
-    def __init__(self,
-                 name: str,
-                 algorithm: AbsAlgorithm,
-                 experience_pool: AbsStore
-                 ):
+    def __init__(self, name: str, algorithm: AbsAlgorithm, experience_pool: AbsStore = None):
```
Comment (ArthurJiang) [inline maro/rl/agent/abs_agent.py:29]: Let's keep ecp_pool

File: maro/rl/agent/abs_agent_manager.py [hunk 3]
```diff
@@ -98,24 +91,16 @@ def choose_action(self, decision_event, snapshot_list):
...
-        agent_id, model_state = self._state_shaper(decision_event, snapshot_list)
-        model_action = self._agent_dict[agent_id].choose_action(
-            model_state, self._explorer.epsilon[agent_id] if self._explorer else None)
-        self._trajectory.append({"state": model_state,
-                                 "action": model_action,
-                                 "reward": None,
-                                 "agent_id": agent_id,
-                                 "event": decision_event})
-        return self._action_shaper(model_action, decision_event, snapshot_list)
+        return NotImplementedError
```
Comment (ArthurJiang) [inline maro/rl/agent/abs_agent_manager.py:94]: add value_based_ag, p_based_ag, ac_based_ag

File: maro/rl/algorithms/torch/dqn.py [hunk 4]
```diff
@@ -32,44 +32,57 @@ def __init__(self, num_actions: int, reward_decay: float, num_training_rounds_pe
...
-        super().__init__(model_dict, optimizer_opt, loss_func_dict, hyper_params)
-        self._train_cnt = 0
+    def __init__(
+        self, eval_model: nn.Module, optimizer_cls, optimizer_params, loss_func, hyper_params: DQNHyperParams,
+        target_model: nn.Module = None
+    ):
+        super().__init__()
         self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self._eval_model = eval_model.to(self._device)
+        self._target_model = clone(eval_model) if target_model is None else target_model
```
Comment (ArthurJiang) [inline maro/rl/algorithms/torch/dqn.py:53]: multi-model use model_dict

File: maro/rl/algorithms/torch/ppo.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        self, num_actions: int, reward_decay: float, clip_ratio: float, policy_train_iters: int,
+        value_train_iters: int, k: int = -1, lamb: float = 1.0
+    ):
+        self.num_actions = num_actions
+        self.reward_decay = reward_decay
+        self.clip_ratio = clip_ratio
+        self.policy_train_iters = policy_train_iters
+        self.value_train_iters = value_train_iters
+        self.k = k
+        self.lamb = lamb
```
Comment (ArthurJiang) [inline maro/rl/algorithms/torch/ppo.py:40]: lam

File: maro/rl/models/torch/mlp_decision_layers.py [hunk 6]
```diff
@@ -16,8 +16,13 @@ class MLPDecisionLayers(nn.Module):
                             hidden layer number, which requires larger than 1.
         output_dim (int): Network output dimension.
         dropout_p (float): Dropout parameter.
+        softmax (bool): If true, the output of the net will be a softmax transformation of the top layer's output.
+            Defaults to False.
     """
-    def __init__(self, *, name: str, input_dim: int, output_dim: int, hidden_dims: [int], dropout_p: float):
+    def __init__(
+        self, *, name: str, input_dim: int, output_dim: int, hidden_dims: [int], dropout_p: float,
+        softmax: bool = False
```
Comment (ArthurJiang) [inline maro/rl/models/torch/mlp_decision_layers.py:24]: enable_softmax, add active_func

File: maro/rl/shaping/k_step_experience_shaper.py [hunk 7]
```diff
@@ -34,28 +35,34 @@ def __init__(self, reward_func: Callable, reward_decay: float, steps: int, is_pe
...
-        return dict(experiences)
+        length = len(trajectory)
+        agent_ids = np.asarray(trajectory.get_by_key["agent_id"])
+        states = np.asarray(trajectory.get_by_key["state"])
+        actions = np.asarray(trajectory.get_by_key["action"])
+        reward_array = np.fromiter(map(self._reward_func, trajectory.get_by_key("metrics")), dtype=np.float32)
+        reward_sums = get_k_step_returns(reward_array, self._reward_decay, k=self._steps)
+        discounts = np.array([self._reward_decay ** min(self._steps, length - i - 1) for i in range(length - 1)])
+        next_states = np.pad(states[self._steps:], (0, length - self._steps - 1), mode="edge")
+        next_actions = np.pad(actions[self._steps:], (0, length - self._steps - 1), mode="edge")
```
Comment (ArthurJiang) [inline maro/rl/shaping/k_step_experience_shaper.py:46]: _step_num

File: maro/rl/shaping/k_step_experience_shaper.py [hunk 8]
```diff
@@ -34,28 +35,34 @@ def __init__(self, reward_func: Callable, reward_decay: float, steps: int, is_pe
...
-                self._reward_decay ** (min(self._steps, len(trajectory) - 1 - i))
-            )
-
-        return dict(experiences)
+        length = len(trajectory)
+        agent_ids = np.asarray(trajectory.get_by_key["agent_id"])
+        states = np.asarray(trajectory.get_by_key["state"])
+        actions = np.asarray(trajectory.get_by_key["action"])
+        reward_array = np.fromiter(map(self._reward_func, trajectory.get_by_key("metrics")), dtype=np.float32)
+        reward_sums = get_k_step_returns(reward_array, self._reward_decay, k=self._steps)
```
Comment (ArthurJiang) [inline maro/rl/shaping/k_step_experience_shaper.py:43]: turncate_reward
get_turncate_reward()

File: maro/rl/storage/column_based_store.py [hunk 9]
```diff
@@ -53,6 +53,11 @@ def __next__(self):
     def __getitem__(self, index: int):
         return {k: lst[index] for k, lst in self._store.items()}
 
+    def __getstate__(self):
+        obj_dict = self.__dict__
```
Comment (ArthurJiang) [inline maro/rl/storage/column_based_store.py:57]: add comments

File: maro/rl/storage/column_based_store.py [hunk 10]
```diff
@@ -75,7 +80,7 @@ def put(self, contents: dict, overwrite_indexes: Sequence = None) -> List[int]:
         """Put new contents in the store.
 
         Args:
-            contents (Sequence): Item object list.
+            contents (dict): Item object list.
```
Comment (ArthurJiang) [inline maro/rl/storage/column_based_store.py:83]: update doc string

File: maro/rl/storage/utils.py [hunk 11]
```diff
@@ -12,9 +12,11 @@ def decorator(func):
         @wraps(func)
         def wrapper(*args, **kwargs):
             contents = args[arg_num]
+            if all(not isinstance(val, list) for val in contents.values()):
+                return func(*args, **kwargs)
             length = len(contents[next(iter(contents))])
-            if any(len(lst) != length for lst in contents.values()):
-                raise ValueError("all sequences in contents should have the same length")
+            if any(not isinstance(val, list) or len(val) != length for val in contents.values()):
```
Comment (ArthurJiang) [inline maro/rl/storage/utils.py:18]: lenght -> col_length

File: maro/rl/storage/utils.py [hunk 12]
```diff
@@ -12,9 +12,11 @@ def decorator(func):
...
+                raise ValueError("values of contents should consist of lists of the same length")
```
Comment (ArthurJiang) [inline maro/rl/storage/utils.py:19]: change to our self-designed exception

File: examples/cim/ac/components/experience_shaper.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        states = np.asarray(trajectory.get_by_key("state"))
+        actions = np.asarray(trajectory.get_by_key("action"))
+        rewards = np.fromiter(
+            map(self._compute_reward, trajectory.get_by_key("event"), [snapshot_list] * len(trajectory)),
+            dtype=np.float32
+        )
+        return {agent_id: {"states": states[agent_ids == agent_id],
+                           "actions": actions[agent_ids == agent_id],
+                           "rewards": rewards[agent_ids == agent_id],
+                           }
```
Comment (ArthurJiang) [inline examples/cim/ac/components/experience_shaper.py:31]: add next_states

File: examples/cim/ac/multi_process_launcher.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+"""
+This script is used to debug distributed algorithm in single host multi-process mode.
+"""
+
+import os
+
+from components.config import config
+
+
+ACTOR_NUM = config.distributed.learner.peer["actor_worker"]  # must be same as in config
```
Comment (ArthurJiang) [inline examples/cim/ac/multi_process_launcher.py:13]: @kaiqli @ysqyang update config with a meaningful name

File: examples/cim/ac/components/experience_shaper.py (hunk 13 above)
Comment (ysqyang) [reply examples/cim/ac/components/experience_shaper.py:31]: The lambda return would require a state series. I think we would be better off passing a state sequence to the algorithm.

**Review by wesley-stone — APPROVED** [review]

**Pull request merged**